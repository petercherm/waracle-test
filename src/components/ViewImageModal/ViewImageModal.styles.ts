import styled, { css, keyframes } from "styled-components";

const overlayFadeInAnimation = keyframes`  
  from { opacity: 0; }
  to { opacity: 1; }
`;

const modalZoomInAnimation = keyframes`  
  from { transform: scale(0.5); }
  to { transform: scale(1); }
`;

const AnimatedOverlay = styled.div`
  animation: ${overlayFadeInAnimation} 0.5s ease-in-out;
`;

const AnimatedModal = styled.div`
  animation: ${modalZoomInAnimation} 0.5s ease-in-out;
`;

export const ViewImageModalOverlay = styled(AnimatedOverlay)(({ theme }) => ({
  display: "flex ",
  justifyContent: "center",
  alignItems: "center",
  backgroundColor: theme.colors.overlayColor,
  position: "fixed",
  top: 0,
  left: 0,
  width: "100vw",
  height: "100vh"
}));

export const ViewImageModalContainer = styled(AnimatedModal)(({ theme }) => ({
  display: "flex",
  flexDirection: "column",
  justifyContent: "center",
  width: "90vh",
  maxHeight: "90vh",
  alignItems: "center",
  backgroundColor: theme.colors.backgroundColor,
  margin: theme.sizes.medium
}));

export const CloseButtonContainer = styled.div(({ theme }) => ({
  display: "flex",
  justifyContent: "flex-end",
  alignSelf: "stretch",
  padding: theme.sizes.medium
}));

export const CloseButton = styled.div.attrs({ role: "button" })(
  ({ theme }) => ({
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    width: theme.sizes.xLarge,
    height: theme.sizes.xLarge,
    backgroundColor: theme.colors.buttonColor,
    borderRadius: "50%",
    cursor: "pointer",
    transition: "background-color 0.5s ease-in-out",

    [":hover"]: {
      backgroundColor: theme.colors.hoverBackgroundColor,

      ["::before"]: {
        color: theme.colors.hightLightColor
      }
    },

    ["::before"]: {
      content: '"X"',
      transition: "color 0.5s ease-in-out"
    }
  })
);

export const ImageContainer = styled.div(({ theme }) => ({
  display: "flex",
  justifyContent: "center",
  alignItems: "center",
  flex: "1 0 auto",
  padding: theme.sizes.medium
}));

export const LargeImage = styled.img({
  width: "100%"
});
