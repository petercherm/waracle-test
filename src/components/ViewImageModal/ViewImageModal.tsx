import {
  BottomStripe,
  CloseButton,
  CloseButtonContainer,
  ImageContainer,
  LargeImage,
  ViewImageModalContainer,
  ViewImageModalOverlay
} from "./ViewImageModal.styles";

export interface ViewImageModalProps {
  imageUrl: string;
  onClose: () => void;
}

export const ViewImageModal = ({ imageUrl, onClose }: ViewImageModalProps) => {
  return (
    <ViewImageModalOverlay data-testid="imageModal__overlay">
      <ViewImageModalContainer>
        <CloseButtonContainer>
          <CloseButton
            onClick={onClose}
            data-testid="imageModal__button--close"
          />
        </CloseButtonContainer>
        <ImageContainer>
          <LargeImage src={imageUrl} data-testid="imageModal__image" />
        </ImageContainer>
        <BottomStripe />
      </ViewImageModalContainer>
    </ViewImageModalOverlay>
  );
};
