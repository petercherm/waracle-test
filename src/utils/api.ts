import { AxiosError } from "axios";

export interface ErrorResponseMessage {
  message: string;
}

export const returnError = (
  error: AxiosError<ErrorResponseMessage, unknown>
) => {
  if (error.response) {
    return {
      error: { message: error.response.data?.message || "Unknown error" }
    };
  }
  return { error };
};
