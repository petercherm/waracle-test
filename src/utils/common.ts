const isObject = (item: any): boolean => {
  return item && typeof item === "object" && !Array.isArray(item);
};

export const mergeDeep = <T, S>(target: T, ...sources: S[]): T => {
  if (!sources.length) return target;
  const source = sources.shift();

  if (isObject(target) && isObject(source)) {
    for (const key in source) {
      if (isObject(source[key])) {
        if (!(target as Record<string, unknown>)[key])
          Object.assign(target, {
            [key]: {}
          });
        mergeDeep(
          (target as Record<string, any>)[key],
          (source as Record<string, any>)[key]
        );
      } else {
        Object.assign(target, {
          [key]: source[key]
        });
      }
    }
  }

  return mergeDeep(target, ...sources);
};
