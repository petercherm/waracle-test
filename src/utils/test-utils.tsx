/* eslint-disable import/export */
import { rootReducer } from "../store/rootReducer";
import { render, RenderResult } from "@testing-library/react";
import { FunctionComponent } from "react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import { ThemeProvider } from "styled-components";
import { baseTheme } from "../theme/baseTheme";

const customRender = (
  ui: JSX.Element,
  {
    initialState = {},
    store = createStore(rootReducer, initialState),
    ...renderOptions
  }: any = {}
): RenderResult => {
  const Wrapper: FunctionComponent = ({ children }) => {
    return (
      <Provider store={store}>
        <ThemeProvider theme={baseTheme}>{children}</ThemeProvider>
      </Provider>
    );
  };
  return render(ui, { wrapper: Wrapper, ...renderOptions });
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
