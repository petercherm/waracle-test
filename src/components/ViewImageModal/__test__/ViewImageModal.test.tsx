import "@testing-library/jest-dom/extend-expect";
import { render } from "../../../utils/test-utils";
import { ViewImageModal, ViewImageModalProps } from "../ViewImageModal";

describe("GIVEN ViewImageModal component", () => {
  const onClose = jest.fn();

  const props: ViewImageModalProps = {
    imageUrl: "/test_url.jpg",
    onClose
  };
  test("it matches the snapshot", () => {
    const container = render(<ViewImageModal {...props} />);

    expect(container).toMatchSnapshot();
  });

  test("it renders all the elements correctly", () => {
    const { queryByTestId } = render(<ViewImageModal {...props} />);

    expect(queryByTestId("imageModal__overlay")).toBeTruthy();
    expect(queryByTestId("imageModal__button--close")).toBeTruthy();
    expect(queryByTestId("imageModal__image")).toBeTruthy();
  });

  test("it calls the onClose() function when the close button is clicked", () => {
    const { getByTestId } = render(<ViewImageModal {...props} />);

    const closeButton = getByTestId("imageModal__button--close");

    closeButton.click();

    expect(onClose).toBeCalled();
  });
});
