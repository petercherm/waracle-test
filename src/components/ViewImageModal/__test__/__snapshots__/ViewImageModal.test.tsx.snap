// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN ViewImageModal component it matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdvvtL sc-dkPtRN iYkkBi kZZWkh"
        data-testid="imageModal__overlay"
      >
        <div
          class="sc-gsDKAQ sc-hKwDye hkdxMk kELOfl"
        >
          <div
            class="sc-eCImPb eJJzMm"
          >
            <div
              class="sc-jRQBWg gGMIlY"
              data-testid="imageModal__button--close"
              role="button"
            />
          </div>
          <div
            class="sc-gKclnd llaENl"
          >
            <img
              class="sc-iCfMLu dhbpeV"
              data-testid="imageModal__image"
              src="/test_url.jpg"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdvvtL sc-dkPtRN iYkkBi kZZWkh"
      data-testid="imageModal__overlay"
    >
      <div
        class="sc-gsDKAQ sc-hKwDye hkdxMk kELOfl"
      >
        <div
          class="sc-eCImPb eJJzMm"
        >
          <div
            class="sc-jRQBWg gGMIlY"
            data-testid="imageModal__button--close"
            role="button"
          />
        </div>
        <div
          class="sc-gKclnd llaENl"
        >
          <img
            class="sc-iCfMLu dhbpeV"
            data-testid="imageModal__image"
            src="/test_url.jpg"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
